### PR TITLE
feat: add OpenID Connect ROPC authenticator and principals

### PR DESCRIPTION
**Pull request by c35sys**
Use Resource Owner Password Credentials Grant: https://tools.ietf.org/html/rfc6749#page-37
Allow use of others principals than the OIDC one
Tested with Keycloak 6.0.1

This change will be reflected in documentation.

### CODE REVIEW

**Comment by c35sys** [comment]
# ROPC Workflow

workflow-signmykey1


1. The user enters signmykey command then enters its credentials, Signmykey client push to Signmykey server the credentials and SSH public key
2. Signmykey server sends user credentials + OIDC client ID and secret to Keycloak using Token Endpoint (Authenticator)
3. Keycloak server responds with an authentication token if successful
4. Signmykey server sends authentication token to Keycloak using Userinfo Endpoint (Principals)
5. Keycloak server responds with user groups as Principals
6. Signmykey server asks Vault to sign the public key with groups as principals
7. Vault gives back the signed key to Signmykey server
8. Signmykey server gives back the signed key to the user

File: builtin/authenticator/oidcropc/Authenticator.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+type oidcToken struct {
+	Token string `json:"access_token"`
+}
+
+// Login method is used to check if a couple of user/password is valid in LDAP.
+func (a *Authenticator) Login(user, password string) (valid bool, swapuser string, err error) {
+
+	payload := strings.NewReader("grant_type=password&" +
```
Comment (pablo-ruth) [inline builtin/authenticator/oidcropc/Authenticator.go]: Using https://golang.org/pkg/net/url/#Values is more idiomatic

File: builtin/authenticator/oidcropc/Authenticator.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+		a.OIDCClientID +
+		"&client_secret=" +
+		a.OIDCClientSecret)
+
+	reqToken, _ := http.NewRequest("POST", a.OIDCTokenEndpoint, payload)
+
+	reqToken.Header.Add("content-type", "application/x-www-form-urlencoded")
+
+	client := http.Client{Timeout: time.Second * 10}
+	resToken, errResToken := client.Do(reqToken)
```
Comment (pablo-ruth) [inline builtin/authenticator/oidcropc/Authenticator.go]: Errors should always be named "err"

File: builtin/authenticator/oidcropc/Authenticator.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	if errResToken != nil {
+		return false, "", errors.New("bad password")
+	}
+
+	defer resToken.Body.Close()
+
+	bodyToken, _ := ioutil.ReadAll(resToken.Body)
```
Comment (pablo-ruth) [inline builtin/authenticator/oidcropc/Authenticator.go]: Check error

File: builtin/authenticator/oidcropc/Authenticator.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+	oidcToken1 := oidcToken{}
+	jsonTokenErr := json.Unmarshal(bodyToken, &oidcToken1)
+
+	log.Debugf("OIDC Token: %s", oidcToken1.Token)
+
+	if jsonTokenErr != nil {
```
Comment (pablo-ruth) [inline builtin/authenticator/oidcropc/Authenticator.go]: Errors should always be named "err"

File: builtin/principals/oidcropc/Principals.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+	p.TransformCase = tc
+
+	return nil
+}
+
+// Get method is used to get the list of principals associated to a specific user.
+func (p Principals) Get(user string) (principals []string, err error) {
+
+	reqInfo, _ := http.NewRequest("GET", p.OIDCUserinfoEndpoint, nil)
```
Comment (pablo-ruth) [inline builtin/principals/oidcropc/Principals.go]: Check error

File: builtin/principals/oidcropc/Principals.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+	// Add HTTP Authorization Header
+	bearer := "Bearer " + user
+	reqInfo.Header.Add("Authorization", bearer)
+
+	client := http.Client{Timeout: time.Second * 10}
+	resInfo, errResInfo := client.Do(reqInfo)
```
Comment (pablo-ruth) [inline builtin/principals/oidcropc/Principals.go]: Errors should always be named "err"

File: builtin/principals/oidcropc/Principals.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+	if errResInfo != nil {
+		log.Fatal(errResInfo)
+	}
+
+	defer resInfo.Body.Close()
+
+	bodyInfo, _ := ioutil.ReadAll(resInfo.Body)
```
Comment (pablo-ruth) [inline builtin/principals/oidcropc/Principals.go]: Check err

File: builtin/principals/oidcropc/Principals.go [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+	// Replace `json:"oidcgroups"` oidcUserinfo struct tag with OIDCUserGroupsEntry config entry
+	bodyInfoChange := []byte(strings.Replace(string(bodyInfo), p.OIDCUserGroupsEntry, "oidcgroups", 1))
+
+	oidcUserinfo1 := oidcUserinfo{}
+	jsonInfoErr := json.Unmarshal(bodyInfoChange, &oidcUserinfo1)
+	if jsonInfoErr != nil {
```
Comment (pablo-ruth) [inline builtin/principals/oidcropc/Principals.go]: Errors should always be named "err"

File: builtin/authenticator/oidcropc/Authenticator.go [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+	reqToken.Header.Add("content-type", "application/x-www-form-urlencoded")
+
+	client := http.Client{Timeout: time.Second * 10}
+	resToken, errResToken := client.Do(reqToken)
+	if errResToken != nil {
+		return false, "", errors.New("bad password")
+	}
+
+	defer resToken.Body.Close()
+
```
Comment (pablo-ruth) [inline builtin/authenticator/oidcropc/Authenticator.go]: We should check http status code for error : https://tools.ietf.org/html/rfc6749#section-5.2

Comment (c35sys) [reply builtin/authenticator/oidcropc/Authenticator.go]: Done

File: builtin/principals/oidcropc/Principals.go (hunk 8 above)
Comment (c35sys) [reply builtin/principals/oidcropc/Principals.go]: Done

File: builtin/principals/oidcropc/Principals.go (hunk 7 above)
Comment (c35sys) [reply builtin/principals/oidcropc/Principals.go]: Done

File: builtin/principals/oidcropc/Principals.go (hunk 6 above)
Comment (c35sys) [reply builtin/principals/oidcropc/Principals.go]: Done

File: builtin/principals/oidcropc/Principals.go (hunk 5 above)
Comment (c35sys) [reply builtin/principals/oidcropc/Principals.go]: Done

File: builtin/authenticator/oidcropc/Authenticator.go (hunk 4 above)
Comment (c35sys) [reply builtin/authenticator/oidcropc/Authenticator.go]: Done

File: builtin/authenticator/oidcropc/Authenticator.go (hunk 3 above)
Comment (c35sys) [reply builtin/authenticator/oidcropc/Authenticator.go]: Done

File: builtin/authenticator/oidcropc/Authenticator.go (hunk 2 above)
Comment (c35sys) [reply builtin/authenticator/oidcropc/Authenticator.go]: Done

File: builtin/authenticator/oidcropc/Authenticator.go (hunk 1 above)
Comment (c35sys) [reply builtin/authenticator/oidcropc/Authenticator.go]: Done

File: builtin/authenticator/oidcropc/Authenticator.go [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+		"username=" +
+		user +
+		"&password=" +
+		password +
+		"&client_id=" +
+		a.OIDCClientID +
+		"&client_secret=" +
+		a.OIDCClientSecret)
+
+	reqToken, _ := http.NewRequest("POST", a.OIDCTokenEndpoint, payload)
```
[the remark this answers is not in the document]
Comment (c35sys) [reply builtin/authenticator/oidcropc/Authenticator.go]: Done

File: builtin/authenticator/oidcropc/Authenticator.go [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+	a.OIDCTokenEndpoint = config.GetString("oidcTokenEndpoint")
+	a.OIDCClientID = config.GetString("oidcClientID")
+	a.OIDCClientSecret = config.GetString("oidcClientSecret")
+	a.OIDCAlternatePrincipal = config.GetBool("oidcAlternatePrincipal")
+
+	return nil
+}
+
+type oidcToken struct {
```
[the remark this answers is not in the document]
Comment (c35sys) [reply builtin/authenticator/oidcropc/Authenticator.go]: Done

File: builtin/principals/oidcropc/Principals.go [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+	return nil
+}
+
+// Get method is used to get the list of principals associated to a specific user.
+func (p Principals) Get(user string) (principals []string, err error) {
+
+	reqInfo, err := http.NewRequest("GET", p.OIDCUserinfoEndpoint, nil)
+	if err != nil {
+		log.Fatal(err)
```
Comment (pablo-ruth) [inline builtin/principals/oidcropc/Principals.go]: log.Fatal does an os.Exit(1) https://golang.org/src/log/log.go#L312 , here we just want to return the error :)

File: builtin/principals/oidcropc/Principals.go [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	// Add HTTP Authorization Header
+	bearer := "Bearer " + user
+	reqInfo.Header.Add("Authorization", bearer)
+
+	client := http.Client{Timeout: time.Second * 10}
+	resInfo, err := client.Do(reqInfo)
+	if err != nil {
+		log.Fatal(err)
```
Comment (pablo-ruth) [inline builtin/principals/oidcropc/Principals.go]: same as line 67

File: builtin/principals/oidcropc/Principals.go [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	defer resInfo.Body.Close()
+
+	bodyInfo, err := ioutil.ReadAll(resInfo.Body)
+	if err != nil {
+		log.Fatal(err)
```
Comment (pablo-ruth) [inline builtin/principals/oidcropc/Principals.go]: same as line 67

File: builtin/principals/oidcropc/Principals.go [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	// Replace `json:"oidcgroups"` oidcUserinfo struct tag with OIDCUserGroupsEntry config entry
+	bodyInfoChange := []byte(strings.Replace(string(bodyInfo), p.OIDCUserGroupsEntry, "oidcgroups", 1))
+
+	oidcUserinfo1 := oidcUserinfo{}
+	err = json.Unmarshal(bodyInfoChange, &oidcUserinfo1)
+	if err != nil {
+		log.Fatal(err)
```
Comment (pablo-ruth) [inline builtin/principals/oidcropc/Principals.go]: same as line 67

File: builtin/authenticator/oidcropc/Authenticator.go (hunk 9 above)
Comment (pablo-ruth) [reply builtin/authenticator/oidcropc/Authenticator.go]: the rfc states that for an "invalid_client" may return a 401, and in general we should exit the function and return an error if the status is not 200

File: builtin/principals/oidcropc/Principals.go (hunk 12 above)
Comment (c35sys) [reply builtin/principals/oidcropc/Principals.go]: Done

File: builtin/principals/oidcropc/Principals.go (hunk 13 above)
Comment (c35sys) [reply builtin/principals/oidcropc/Principals.go]: Done

File: builtin/principals/oidcropc/Principals.go (hunk 15 above)
Comment (c35sys) [reply builtin/principals/oidcropc/Principals.go]: Done

File: builtin/principals/oidcropc/Principals.go (hunk 14 above)
Comment (c35sys) [reply builtin/principals/oidcropc/Principals.go]: Done

File: builtin/authenticator/oidcropc/Authenticator.go (hunk 9 above)
Comment (c35sys) [reply builtin/authenticator/oidcropc/Authenticator.go]: Done

File: builtin/principals/oidcropc/Principals.go (hunk 15 above)
Comment (pablo-ruth) [reply builtin/principals/oidcropc/Principals.go]: still one at line 93

Comment (c35sys) [reply builtin/principals/oidcropc/Principals.go]: Done

**Pull request merged**